components/modules/ble.c

```diff
@@ -0,0 +1,1062 @@
+
+/**
+ * Utility function to log an array of bytes.
+ */
+static void
+print_addr(const void *addr)
+{
+    const uint8_t *u8p;
+
+    u8p = addr;
+    MODLOG_DFLT(INFO, "%02x:%02x:%02x:%02x:%02x:%02x",
+                u8p[5], u8p[4], u8p[3], u8p[2], u8p[1], u8p[0]);
+}
+
+static int
+gethexval(char c) {
+  if (c < '0') {
+    return -1;
+  }
+  if (c <= '9') {
+    return c - '0';
+  }
+  if (c < 'a') {
+    c += 'a' - 'A';
+  }
+  if (c < 'a') {
+    return -1;
+  }
+  if (c <= 'f') {
+    return c - 'a' + 10;
+  }
+  return -1;
+}
+
+static int
+decodehex(const char *s) {
+  // two characters
+  int v1 = gethexval(s[0]);
+  int v2 = gethexval(s[1]);
+
+  if (v1 < 0 || v2 < 0) {
+    return -1;
+  }
+  return (v1 << 4) + v2;
+}
+
+static bool
+convert_uuid(ble_uuid_any_t *uuid, const char *s) {
+  int len = strlen(s);
+  char decodebuf[32];
+
+  const char *sptr = s + len;
+  int i;
+
+  for (i = 0; sptr > s && i < sizeof(decodebuf); ) {
+    if (sptr < s + 2) {
+      return false;
+    }
+    if (sptr[-1] == '-') {
+      sptr--;
+      continue;
+    }
+    sptr -= 2;
+
+    int val = decodehex(sptr);
+    if (val < 0) {
+      return false;
+    }
+    decodebuf[i++] = val;
+  }
+
+  if (sptr != s) {
+    return false;
+  }
+
+  if (ble_uuid_init_from_buf(uuid, decodebuf, i) != 0) {
+    return false;
+  }
+
+  return true;
+}
+
+static void
+free_gatt_svcs(lua_State *L, const struct ble_gatt_svc_def * svcs) {
+  void *tofree = (void *) svcs;
+  // Need to unref anything
+  if (!svcs) {
+    return;
+  }
+
+  for (; svcs->characteristics; svcs++) {
+    const struct ble_gatt_chr_def *chrs = svcs->characteristics;
+    for (; chrs->arg; chrs++) {
+      luaL_unref(L, LUA_REGISTRYINDEX, (int) chrs->arg);
+    }
+  }
+
+  free(tofree);
+}
+
+static int
+lble_access_cb(uint16_t conn_handle, uint16_t attr_handle, struct ble_gatt_access_ctxt *ctxt, void *arg) {
+  // Actually the only thing we care about is the arg and the ctxt
+
+  size_t task_block_size = sizeof(task_block_t);
+
+  if (ctxt->op == BLE_GATT_ACCESS_OP_WRITE_CHR) {
+    task_block_size += OS_MBUF_PKTLEN(ctxt->om);
+  }
+
+  task_block_t *task_block = malloc(task_block_size);
+  if (!task_block) {
+    return BLE_ATT_ERR_UNLIKELY;
+  }
+
+  memset(task_block, 0, sizeof(*task_block));
+
+  task_block->ctxt = ctxt;
+  task_block->arg = arg;
+  task_block->seqno = seqno++;
+  task_block->buffer = (char *) (task_block + 1);
+
+  if (ctxt->op == BLE_GATT_ACCESS_OP_WRITE_CHR) {
+    task_block->length = OS_MBUF_PKTLEN(ctxt->om);
+    uint16_t outlen;
+    if (ble_hs_mbuf_to_flat(ctxt->om, task_block->buffer, task_block->length, &outlen)) {
+      return BLE_ATT_ERR_UNLIKELY;
+    }
+  }
+
+  if (!task_post(TASK_PRIORITY_HIGH, task_handle, (task_param_t) task_block)) {
+    free(task_block);
+    return BLE_ATT_ERR_UNLIKELY;
+  }
+
+  response_message_t message;
+
+  while (1) {
+    if (xQueueReceive(response_queue, &message, (TickType_t) (10000/portTICK_PERIOD_MS) ) != pdPASS) {
```

Comment:
I'm not across Bluetooth very much, but is it safe to have such a long timeout here, in what I presume to be the Bluetooth thread?

Reply:
I'm going to reduce this to 2 seconds....